taskfiles/toolchains.yaml

```
@@ -1,13 +1,76 @@
 version: "3"
 
+includes:
+  utils: "../tools/yscope-dev-utils/exports/taskfiles/utils/utils.yaml"
+
+vars:
+  G_RUST_TOOLCHAIN_ROOT: "{{.G_BUILD_DIR}}/toolchains/rust"
+
+  # Rust toolchain installation paths
+  G_CARGO_HOME: "{{.G_RUST_TOOLCHAIN_ROOT}}/.cargo"
+  G_RUSTUP_HOME: "{{.G_RUST_TOOLCHAIN_ROOT}}/.rustup"
+
+  # Rust toolchain checksum file
+  G_RUST_TOOLCHAIN_CHECKSUM_FILE: "{{.G_BUILD_DIR}}/toolchains-rust.md5"
+
+  G_RUST_TOOLCHAIN_ENV_FILE: "{{.G_RUST_TOOLCHAIN_ROOT}}/env"
+
 tasks:
   rust:
     run: "once"
+    env:
+      CARGO_HOME: "{{.G_CARGO_HOME}}"
+      RUSTUP_HOME: "{{.G_RUSTUP_HOME}}"
+    vars:
+      CHECKSUM_FILE: "{{.G_RUST_TOOLCHAIN_CHECKSUM_FILE}}"
+      # The checksum only tracks the environment files and settings to avoid unnecessary
+      # re-execution when CARGO_HOME/RUSTUP_HOME cache contents change (e.g., downloaded crates,
+      # toolchain auto-updates). The task will only re-run when the toolchain configuration itself
+      # changes.
+      OUTPUT_PATTERNS:
+        - "{{.G_CARGO_HOME}}/env"
+        - "{{.G_RUSTUP_HOME}}/settings.toml"
+        - "{{.G_RUST_TOOLCHAIN_ENV_FILE}}"
+    sources:
+      - "{{.ROOT_DIR}}/taskfile.yaml"
+      - "{{.TASKFILE}}"
+    generates: ["{{.CHECKSUM_FILE}}"]
+    deps:
+      - task: "utils:checksum:validate"
+        vars:
+          CHECKSUM_FILE: "{{.CHECKSUM_FILE}}"
+          INCLUDE_PATTERNS:
+            ref: ".OUTPUT_PATTERNS"
     cmds:
-      # Installs Rust toolchains using rustup. If the toolchains are already installed, this task
-      # will ensure they are updated to the latest versions.
       - |-
         curl --proto "=https" --tlsv1.2 --silent --show-error --fail https://sh.rustup.rs \
           | sh -s -- --default-toolchain none -y
-        . "$HOME/.cargo/env"
+
+      # Create an environment file to configure the custom Rust toolchain location.
+      # Source this file before using cargo/rustup commands to ensure they use the toolchain
+      # installed in `{{.G_RUST_TOOLCHAIN_ROOT}}`.
+      # For details, check:
+      # https://rust-lang.github.io/rustup/installation/index.html#choosing-where-to-install
+      - |-
+        cat > {{.G_RUST_TOOLCHAIN_ENV_FILE}} << 'EOF'
+        #!/bin/sh
+        export CARGO_HOME={{.G_CARGO_HOME}}
+        export RUSTUP_HOME={{.G_RUSTUP_HOME}}
+        . {{.G_CARGO_HOME}}/env
+        EOF
+        chmod +x {{.G_RUST_TOOLCHAIN_ENV_FILE}}
+
+      - |-
+        . {{.G_RUST_TOOLCHAIN_ENV_FILE}}
         rustup toolchain install nightly --allow-downgrade --component clippy,rustfmt
+        rustup toolchain install stable
+        rustup default stable
```

i haven't tried - if we install stable first, would it be the default toolchain?

```suggestion
        rustup toolchain install stable
        rustup toolchain install nightly --allow-downgrade --component clippy,rustfmt
```